PsimagLite/src/ArnoldiIteration.hh

```diff
@@ -0,0 +1,183 @@
+ * only the A*v operation is needed
+ */
+template <typename MatrixType_>
+class ArnoldiIteration {
+public:
+
+	using MatrixType = MatrixType_;
+	using ComplexOrRealType = typename MatrixType_::value_type;
+	using RealType = typename Real<ComplexOrRealType>::Type;
+	using ParamsType = ParametersForSolver<RealType>;
+	using DenseMatrixType = Matrix<ComplexOrRealType>;
+	using VectorType = std::vector<ComplexOrRealType>;
+
+	ArnoldiIteration(const ParamsType& params)
+	    : params_(params)
+	{
+	}
+
+	/**
+	 * \brief This function decomposes a into h = q^* a q
+	 *
+	 * Computes a basis of the (n + 1)-Krylov subspace of the matrix A.
+	 * This is the space spanned by the vectors {b, Ab, ..., A^n b},
+	 * where b is the initial vector
+	 *
+	 * \param[out] q The n + 1 vectors of the orthonormal basis
+	 *               of the Krylov subspace; each of size m
+	 *
+	 * \param[out] h  The (n + 1) x n  upper Hessenberg on basis Q.
+	 *
+	 * \param[in] a The matrix to be decomposed
+	 *
+	 * \param[in] init_vector The initial vector for Krylov
+	 *
+	 * \param[in] n The number of Krylov vectors
+	 */
+	int decompose(std::vector<VectorType>& q, DenseMatrixType& h, const MatrixType& a, const VectorType& init_vector, unsigned int n)
+	{
+		std::string msg = "ArnoldiIteration::decompose():";
+		SizeType m = a.rows();
+		if (m != a.cols()) {
+			throw std::runtime_error(msg + " A not square\n");
+		}
+
+		if (n < 2) {
+			throw std::runtime_error(msg + " n must be greater than one\n");
+		}
+
+		h.resize(n, n);
+		h.setTo(0.);
+		VectorType zero(m);
+		q.resize(n, zero);
+
+		double eps = params_.tolerance;
+		int error_ret = 0;
+
+		// We do not normalize or touch the init_vector
+		q[0] = init_vector;
+
+		double denom = 0; // used as error measure
+		VectorType v(m);
+		for (SizeType k = 0; k < n; ++k) {
+			std::fill(v.begin(), v.end(), 0.);
+
+			a.matrixVectorProduct(v, q[k]); // v += A|q_{k}>
+
+			SizeType total = std::min(n, static_cast<unsigned int>(k + 1));
+			for (SizeType j = 0; j < total; ++j) {
+				h(j, k) = scalarProduct(q[j], v);
+				for (SizeType jj = 0; jj < m; ++jj) {
+					v[jj] -= h(j, k) * q[j][jj];
+				}
+			}
+
+			// Make h square so we can diagonalize
+			if (k + 1 == n) {
+				break;
+			}
+
+			h(k + 1, k) = PsimagLite::norm(v);
+			denom = PsimagLite::real(h(k + 1, k));
+			if (eps > 0) { // User want early exit if possible (recommended)
+				if (denom > eps) {
+					double factor = 1. / denom;
+					q[k + 1] <= factor* v;
+				}
+				else {
+					q.resize(k + 1);
+					h.resize(k + 1, k + 1);
+					break;
+				}
+			}
+			else { // Do all iterations, no early exit
+				bool is_zero = (denom == 0.);
+				if (is_zero && error_ret == 0) {
+					error_ret = k; // k is assured to be greater than zero
+				}
+
+				double factor = (is_zero) ? 1. : 1. / denom;
+				q[k + 1] <= factor* v;
+			}
+		}
+
+		std::cout << "ArnolidIteration: done after " << q.size() << " steps;";
+		std::cout << " error=" << denom << "\n";
```

Comment:
Did you mean to commit this?
Should the printing to the standard output only happen in verbose/debug mode?